### PR TITLE
layer API

### PR DESCRIPTION
**Pull request by ChristianBeilschmidt**

### CODE REVIEW

**Comment by ChristianBeilschmidt** [comment]
[[RESPECT] 3.5.1 Admin-Schnittstelle für Rasterdaten: Layer](https://trello.com/c/yodxVuAI/1082-respect-351-admin-schnittstelle-f%C3%BCr-rasterdaten-layer)

File: geoengine/layers.py [hunk 1]
```diff
@@ -0,0 +1,660 @@
...
+
+        assert False, 'Invalid listing type'
+
+
+class LayerCollection:
+    '''A layer collection'''
+
+    name: str
+    description: str
+    collection_id: LayerCollectionId
```
Comment (michaelmattig) [inline geoengine/layers.py:120]: maybe group the two id parts together in a new type, like in the backend

File: geoengine/layers.py [hunk 2]
```diff
@@ -0,0 +1,660 @@
...
+                  timeout: int = 60) -> LayerId:
+        '''Add a layer to this collection'''
+        # pylint: disable=too-many-arguments
+
+        if self.provider_id != LAYER_DB_PROVIDER_ID:
+            raise ModificationNotOnLayerDbException('Layer collection is not stored in the layer database')
+
+        layer_id = _add_layer_to_collection(name, description, workflow, symbology, self.collection_id, timeout)
+
+        self.items.append(LayerCollectionListing(
```
Comment (michaelmattig) [inline geoengine/layers.py:272]: is that a good idea? maybe due to pagination and sorting it wouldn't directly be visible? Maybe reload the items? Maybe differentiate between a Collection and a CollectionPage?

File: geoengine/layers.py [hunk 3]
```diff
@@ -0,0 +1,660 @@
...
+
+@dataclass(repr=False)
+class LayerCollectionListing:
+    '''A layer collection listing as item of a collection'''
+
+    listing_id: Union[LayerId, LayerCollectionId]
+    provider_id: LayerProviderId
+    name: str
+    description: str
+    type: LayerCollectionListingType
```
Comment (michaelmattig) [inline geoengine/layers.py:67]: why not use a subclass for each type? That would also help to get rid of any `if type == ...` statements in the code

File: geoengine/layers.py [hunk 4]
```diff
@@ -0,0 +1,660 @@
...
+        if self.type == LayerCollectionListingType.COLLECTION:
+            return layer_collection(cast(LayerCollectionId, self.listing_id), self.provider_id, offset, limit, timeout)
+
+        if self.type == LayerCollectionListingType.LAYER:
+            return layer(cast(LayerId, self.listing_id), self.provider_id, timeout)
+
+        assert False, 'Invalid listing type'
+
+
+class LayerCollection:
```
Comment (michaelmattig) [inline geoengine/layers.py:115]: maybe store offset & limit?

File: geoengine/auth.py [hunk 5]
```diff
@@ -37,9 +37,15 @@ class Session:
...
+    __admin_token: Optional[UUID] = None
+
     session: ClassVar[Optional[Session]] = None
 
-    def __init__(self, server_url: str, credentials: Tuple[str, str] = None, token: str = None) -> None:
+    def __init__(self,
+                 server_url: str,
+                 credentials: Optional[Tuple[str, str]] = None,
+                 token: Optional[str] = None,
+                 admin_token: Optional[str] = None) -> None:
```
Comment (michaelmattig) [inline geoengine/auth.py:48]: what about loading from env?
add to the pydoc?

File: geoengine/layers.py [hunk 6]
```diff
@@ -0,0 +1,660 @@
...
+
+    def __repr__(self) -> str:
+        '''String representation of a `LayerCollectionListing`'''
+
+        buf = StringIO()
+
+        buf.write(f"name: {self.name}{os.linesep}")
+        buf.write(f"description: {self.description}{os.linesep}")
+        buf.write(f"id: {self.listing_id}{os.linesep}")
+        buf.write(f"provider id: {self.provider_id}{os.linesep}")
```
Comment (michaelmattig) [inline geoengine/layers.py:77]: add the type

File: geoengine/layers.py [hunk 7]
```diff
@@ -0,0 +1,660 @@
...
+        '''HTML representation for Jupyter notebooks'''
+
+        buf = StringIO()
+
+        buf.write('<table>')
+        buf.write(f"<tr><th>name</th><td>{self.name}</td></tr>")
+        buf.write(f"<tr><th>description</th><td>{self.description}</td></tr>")
+        buf.write(f"<tr><th>id</th><td>{self.listing_id}</td></tr>")
+        buf.write(f"<tr><th>provider id</th><td>{self.provider_id}</td></tr>")
+        buf.write('</table>')
```
Comment (michaelmattig) [inline geoengine/layers.py:91]: add the type

File: geoengine/layers.py [hunk 8]
```diff
@@ -0,0 +1,660 @@
...
+def _delete_layer_from_collection(collection_id: LayerCollectionId,
+                                  layer_id: LayerId,
+                                  timeout: int = 60) -> None:
+    '''Delete a layer from a collection'''
+
+    session = get_session()
+
+    response = req.delete(
+        f'{session.server_url}/layerDb/collections/{collection_id}/layers/{layer_id}',
+        headers=session.admin_auth_header,
```
Comment (michaelmattig) [inline geoengine/layers.py:532]: maybe throw an error if it is not set instead of doing the request? also at the other places.

**Review by michaelmattig — CHANGES_REQUESTED** [review]

File: geoengine/layers.py (hunk 8 above)
Comment (ChristianBeilschmidt) [reply geoengine/layers.py:532]: The property `admin_auth_header` throws an error if it is unset.

File: geoengine/layers.py (hunk 1 above)
Comment (ChristianBeilschmidt) [reply geoengine/layers.py:120]: This adds no advantage currently, so I would leave it as us until we rethink our layer stuff and if this all makes sense.

File: geoengine/layers.py (hunk 3 above)
Comment (ChristianBeilschmidt) [reply geoengine/layers.py:67]: They are no subclasses. You can make isinstance in a later version of Python, when newtypes are actual types.

File: geoengine/auth.py (hunk 5 above)
Comment (ChristianBeilschmidt) [reply geoengine/auth.py:48]: [f2b4d39](https://github.com/geo-engine/geoengine-python/pull/83/commits/f2b4d398bf04c1f091a10d1550d9962b65a91035)

File: geoengine/layers.py (hunk 7 above)
Comment (ChristianBeilschmidt) [reply geoengine/layers.py:95]: [c6d8879](https://github.com/geo-engine/geoengine-python/pull/83/commits/c6d8879dbbde1ba5110530339e524e9f1e3a9ef2)

File: geoengine/layers.py (hunk 4 above)
Comment (ChristianBeilschmidt) [reply geoengine/layers.py:127]: Now, we load everything:
[c6d8879](https://github.com/geo-engine/geoengine-python/pull/83/commits/c6d8879dbbde1ba5110530339e524e9f1e3a9ef2)

File: geoengine/layers.py (hunk 2 above)
Comment (ChristianBeilschmidt) [reply geoengine/layers.py:289]: For now, we load everything until we have more experience with the API:
[c6d8879](https://github.com/geo-engine/geoengine-python/pull/83/commits/c6d8879dbbde1ba5110530339e524e9f1e3a9ef2)

File: geoengine/layers.py [hunk 9]
```diff
@@ -486,16 +501,31 @@ def layer_collection(layer_collection_id: Optional[LayerCollectionId] = None,
...
+            timeout=timeout,
+        )
 
-    return LayerCollection.from_response(response.json())
+        if not response.ok:
+            raise GeoEngineException(response.json())
+
+        page: LayerCollectionResponse = response.json()
+
+        if len(page['items']) == 0:
```
Comment (michaelmattig) [inline geoengine/layers.py:520]: you can already stop requesting once fewer than `limit` elements are returned

File: geoengine/layers.py (hunk 3 above)
Comment (michaelmattig) [reply geoengine/layers.py:67]: There is inheritance aka subclasses in python or what am I missing? https://www.w3schools.com/python/python_inheritance.asp

File: geoengine/layers.py [hunk 10]
```diff
@@ -0,0 +1,687 @@
...
+def _add_layer_to_collection(name: str,
+                             description: str,
+                             workflow: Dict[str, Any],  # TODO: improve type
+                             symbology: Optional[Dict[str, Any]],  # TODO: improve type
+                             collection_id: LayerCollectionId,
+                             timeout: int = 60) -> LayerId:
+    '''Add a new layer'''
+    # pylint: disable=too-many-arguments
+
+    session = get_session()
```
Comment (michaelmattig) [inline geoengine/layers.py:653]: has to be adjusted to new route

Comment (ChristianBeilschmidt) [reply geoengine/layers.py:653]: I already did that, it is `f'{session.server_url}/layerDb/collections/{collection_id}/layers',`

File: geoengine/layers.py (hunk 9 above)
Comment (ChristianBeilschmidt) [reply geoengine/layers.py]: [c6694f3](https://github.com/geo-engine/geoengine-python/pull/83/commits/c6694f39632bea3bcd78a755615d28851e37daf9)

**Review by michaelmattig — APPROVED** [review]

**Review by michaelmattig — APPROVED** [review]

**Comment by michaelmattig** [comment]
bors merge

**Comment by michaelmattig** [comment]
bors merge

**Pull request merged**